### PR TITLE
feat: can now specify a condition on merge insert's 'when matched' behavior

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -112,16 +112,20 @@ def execute(self, data_obj: ReaderLike, *, schema: Optional[pa.Schema] = None):
...
         Configure the operation to update matched rows
 
         After this method is called, when the merge insert operation executes,
         any rows that match both the source table and the target table will be
         updated.  The rows from the target table will be removed and the rows
         from the source table will be added.
+
+        An optional condition may be specified.  This should be an SQL filter
+        and, if present, then only rows matching rows that also match this
+        filter will be updated.
```
Comment (wjones127) [inline python/python/lance/dataset.py:126]: Question: if a row does match this predicate, is it now part of the "when_not_matched" group?

File: python/python/tests/test_dataset.py [hunk 2]
```diff
@@ -898,6 +912,38 @@ def test_merge_insert(tmp_path: Path):
         dataset.merge_insert("a").execute(new_table)
 
 
+def test_merge_insert_conditional_upsert_example(tmp_path: Path):
```
Comment (wjones127) [inline python/python/tests/test_dataset.py:915]: This has conditional update, but I wonder about what happens when there are also inserts? (`when_not_matched_insert_all`)

File: python/python/lance/dataset.py [hunk 3]
```diff
@@ -112,16 +112,20 @@ def execute(self, data_obj: ReaderLike, *, schema: Optional[pa.Schema] = None):
...
+    def when_matched_update_all(self, condition: Optional[str] = None):
         """
         Configure the operation to update matched rows
 
         After this method is called, when the merge insert operation executes,
         any rows that match both the source table and the target table will be
         updated.  The rows from the target table will be removed and the rows
         from the source table will be added.
+
+        An optional condition may be specified.  This should be an SQL filter
```
Comment (wjones127) [inline python/python/lance/dataset.py:124]: Also, should we mention to use the `target.` and `source.` prefixes in column references?

File: python/python/tests/test_dataset.py (hunk 2 above)
Comment (westonpace) [reply python/python/tests/test_dataset.py:915]: There is a test for this in the rust unit tests (and I think in the python unit test above).  This unit test is actually superfluous but I wanted to make sure the exact scenario that the user was asking for worked.

File: python/python/lance/dataset.py (hunk 1 above)
Comment (westonpace) [reply python/python/lance/dataset.py:126]: It is not.  I will update the docstring to clarify this.

File: python/python/lance/dataset.py (hunk 3 above)
Comment (westonpace) [reply python/python/lance/dataset.py:124]: Definitely.  I will add that.

File: python/python/lance/dataset.py (hunk 1 above)
Comment (westonpace) [reply python/python/lance/dataset.py]: I have clarified the docstring here.

File: python/python/lance/dataset.py (hunk 3 above)
Comment (westonpace) [reply python/python/lance/dataset.py:124]: Added

**Pull request merged**